### PR TITLE
Allow deprecating input fields and arguments

### LINKED ISSUE #110: Allow deprecation of input values (arguments and input fields)
I haven't find anything about it in gql specs. Am I missing something?

```php
private static function printInputObject(InputObjectType $type)
    {
        $fields = array_values($type->getFields());
        return self::printDescription($type) . 
            "input {$type->name} {\n" .
                implode("\n", array_map(function($f, $i) {
                    return self::printDescription($f, '  ', !$i) . '  ' . self::printInputValue($f);
                }, $fields, array_keys($fields))) . "\n" .
            "}";
    }
```
doesn't print deprecation reason vs
```php
private static function printObject(ObjectType $type)
    {
        $interfaces = $type->getInterfaces();
        $implementedInterfaces = !empty($interfaces) ?
            ' implements ' . implode(', ', array_map(function($i) {
                return $i->name;
            }, $interfaces)) : '';
        return self::printDescription($type) .
            "type {$type->name}$implementedInterfaces {\n" .
                self::printFields($type) . "\n" .
            "}";
    }

private static function printFields($type)
    {
        $fields = array_values($type->getFields());
        return implode("\n", array_map(function($f, $i) {
                return self::printDescription($f, '  ', !$i) . '  ' .
                    $f->name . self::printArgs($f->args, '  ') . ': ' .
                    (string) $f->getType() . self::printDeprecated($f);
            }, $fields, array_keys($fields)));
    }
```

Thanks!

### PR DESCRIPTION
**Pull request by TaProhm**
Resolves #110

### CODE REVIEW

**Comment by ruudk** [comment]
@TaProhm thanks so much for this 🙏🏻

File: src/Type/Introspection.php [hunk 1]
```diff
@@ -392,9 +394,31 @@ static function (EnumValueDefinition $value): bool {
...
+                        if ($type instanceof InputObjectType) {
+                            $fields = $type->getFields();
+
+                            if (! ($args['includeDeprecated'] ?? false)) {
+                                return \array_filter(
+                                    $fields,
+                                    static function (InputObjectField $field): bool {
+                                        return $field->deprecationReason === null
+                                            || $field->deprecationReason === '';
+                                    }
```
Comment (spawnia) [inline src/Type/Introspection.php:413]:
```suggestion
                                    static fn (InputObjectField $field): bool => $field->deprecationReason === null
                                            || $field->deprecationReason === '',
```

File: src/Type/Introspection.php [hunk 2]
```diff
@@ -469,7 +493,27 @@ public static function _field(): ObjectType
...
+                    'resolve' => static function (FieldDefinition $field, $args): array {
+                        $values = $field->args;
+
+                        if (! ($args['includeDeprecated'] ?? false)) {
+                            return \array_filter(
+                                $values,
+                                static function (Argument $value): bool {
+                                    return $value->deprecationReason === null
+                                        || $value->deprecationReason === '';
+                                }
```
Comment (spawnia) [inline src/Type/Introspection.php:511]:
```suggestion
                                static fn (Argument $value): bool => $value->deprecationReason === null
                                        || $value->deprecationReason === '',
```

File: tests/Type/DefinitionTest.php [hunk 3]
```diff
@@ -920,13 +920,19 @@ public function testAcceptsAnObjectTypeWithFieldArgs(): void
...
+                        ],
                     ],
                 ],
             ],
         ]);
         $objType->assertValid();
-        self::assertDidNotCrash();
+        $argument = $objType->getField('goodField')->getArg('goodArg');
+        self::assertInstanceOf(Argument::class, $argument);
+        self::assertEquals(true, $argument->isDeprecated());
```
Comment (spawnia) [inline tests/Type/DefinitionTest.php:934]:
```suggestion
        self::assertTrue($argument->isDeprecated());
```

File: tests/Type/DefinitionTest.php [hunk 4]
```diff
@@ -1436,12 +1442,16 @@ public function testAcceptsAnInputObjectTypeWithFields(): void
...
+                    'deprecationReason' => 'Just because',
                 ],
             ],
         ]);
 
         $inputObjType->assertValid();
-        self::assertSame(Type::string(), $inputObjType->getField($fieldName)->getType());
+        $field = $inputObjType->getField($fieldName);
+        self::assertSame(Type::string(), $field->getType());
+        self::assertEquals(true, $field->isDeprecated());
```
Comment (spawnia) [inline tests/Type/DefinitionTest.php:1453]:
```suggestion
        self::assertTrue($field->isDeprecated());
```

File: tests/Type/DefinitionTest.php [hunk 5]
```diff
@@ -1453,12 +1463,16 @@ public function testAcceptsAnInputObjectTypeWithAFieldFunction(): void
...
                     'type' => Type::string(),
+                    'deprecationReason' => 'Just because',
                 ],
             ],
         ]);
 
         $inputObjType->assertValid();
+        $field = $inputObjType->getField($fieldName);
         self::assertSame(Type::string(), $inputObjType->getField($fieldName)->getType());
+        self::assertEquals(true, $field->isDeprecated());
```
Comment (spawnia) [inline tests/Type/DefinitionTest.php:1474]:
```suggestion
        self::assertTrue($field->isDeprecated());
```

File: tests/Type/DirectiveTest.php [hunk 6]
```diff
@@ -28,5 +32,56 @@ public function testDefinesADirectiveWithNoArgs(): void
...
+            [
+                'name' => 'foo',
+                'description' => null,
+                'deprecationReason' => 'Just because',
+                'defaultValue' => null,
+                'astNode' => null,
+            ],
+            (array) $argumentFoo
+        );
+        self::assertEquals(true, $argumentFoo->isDeprecated());
```
Comment (spawnia) [inline tests/Type/DirectiveTest.php:70]:
```suggestion
        self::assertTrue($argumentFoo->isDeprecated());
```

File: tests/Type/DirectiveTest.php [hunk 7]
```diff
@@ -28,5 +32,56 @@ public function testDefinesADirectiveWithNoArgs(): void
...
+            [
+                'name' => 'bar',
+                'description' => null,
+                'deprecationReason' => null,
+                'defaultValue' => null,
+                'astNode' => null,
+            ],
+            (array) $argumentBar
+        );
+        self::assertEquals(false, $argumentBar->isDeprecated());
```
Comment (spawnia) [inline tests/Type/DirectiveTest.php:83]:
```suggestion
        self::assertFalse($argumentBar->isDeprecated());
```

**Review by spawnia — COMMENTED** [review]
Nice work!

**Comment by TaProhm** [comment]
@spawnia Thanks for your review, I've made corrections according to your suggestions and fixed the failing test.

File: src/Type/Introspection.php [hunk 8]
```diff
@@ -392,9 +394,31 @@ static function (EnumValueDefinition $value): bool {
...
+                        if ($type instanceof InputObjectType) {
+                            $fields = $type->getFields();
+
+                            if (! ($args['includeDeprecated'] ?? false)) {
+                                return \array_filter(
+                                    $fields,
+                                    static fn (InputObjectField $field): bool =>
+                                        $field->deprecationReason === null
+                                        || $field->deprecationReason === '',
+
```
Comment (spawnia) [inline src/Type/Introspection.php:413]:
```suggestion
```

File: src/Type/Introspection.php [hunk 9]
```diff
@@ -469,7 +493,27 @@ public static function _field(): ObjectType
...
+                    'resolve' => static function (FieldDefinition $field, $args): array {
+                        $values = $field->args;
+
+                        if (! ($args['includeDeprecated'] ?? false)) {
+                            return \array_filter(
+                                $values,
+                                static fn (Argument $value): bool =>
+                                    $value->deprecationReason === null
+                                    || $value->deprecationReason === '',
+
```
Comment (spawnia) [inline src/Type/Introspection.php:511]:
```suggestion
```

**Pull request merged**